### PR TITLE
WIP - Adding npm package versioning

### PR DESCRIPTION
**Pull request by Katka92**
### What does this PR do?
Add npm package versioning.

### What issues does this PR fix or reference?
Publish che-e2e npm package into public registry https://github.com/eclipse/che/issues/18772
Have proper versioning of e2e npm package https://github.com/eclipse/che/issues/18771

### How to test this PR?
To verify that dependencies are not broken, run any test.
To verify package version setup, first set env var `export NPM_AUTH_TOKEN=aaa` and  run `mvn install`.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by dmytro-ndp** [comment]
@Katka92: IMHO, commands in `tests/e2e/README.md` are needed to be updated as well.

File: tests/e2e/publish.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    echo "Building and pushing latest E2E package version $CHE_VERSION"
+    VERSION=${CHE_VERSION%"-SNAPSHOT"}-$TIME
+    TAG="nightly"
+else 
+    echo "Building and pushing E2E package release version $CHE_VERSION"
+    VERSION=$CHE_VERSION-$TIME
+    TAG=$CHE_VERSION
+fi
+
+npm version $VERSION
```
Comment (dmytro-ndp) [inline tests/e2e/publish.sh:15]: Can't find `$VERSION` declaration.
Shouldn't we use `$TAG` instead?

Comment (dmytro-ndp) [reply tests/e2e/publish.sh:15]: Do we really need `$TAG` in there?

File: tests/e2e/pom.xml [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                <executions>
+                    <execution>
+                        <id>remove-old-resources</id>
+                        <phase>clean</phase>
+                        <goals>
+                            <goal>run</goal>
+                        </goals>
+                        <configuration>
+                            <target>
+                                <exec dir="${basedir}" executable="docker" failonerror="true">
```
Comment (benoitf) [inline tests/e2e/pom.xml]: do we need docker to remove files ?

nodejs:10 is also EOL

File: tests/e2e/pom.xml [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                        <id>install-build-lint</id>
+                        <phase>compile</phase>
+                        <goals>
+                            <goal>run</goal>
+                        </goals>
+                        <configuration>
+                            <target>
+                                <!-- build tests and test them for lint -->
+                                <exec dir="${basedir}" executable="docker" failonerror="true">
+                                    <arg value="run" />
```
Comment (benoitf) [inline tests/e2e/pom.xml:61]: specify `--user` with current user id if it's for file permissions issue

File: tests/e2e/publish.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+TIME=$(date +%s)
+
+if [[ $CHE_VERSION == *SNAPSHOT ]] ; then 
+    echo "Building and pushing latest E2E package version $CHE_VERSION"
+    VERSION=${CHE_VERSION%"-SNAPSHOT"}-$TIME
+    TAG="nightly"
```
Comment (dmytro-ndp) [inline tests/e2e/publish.sh]: "TAG" variable looks redundant now.

File: tests/e2e/package-lock.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "e2e",
-  "version": "1.0.0",
+  "version": "7.29.0-1617798294",
```
Comment (dmytro-ndp) [inline tests/e2e/package-lock.json]: Do we need specific Che version in here?

**Comment by dmytro-ndp** [comment]
[crw-ci-test --rebuild]

File: tests/e2e/pom.xml (hunk 3 above)
Comment (Katka92) [reply tests/e2e/pom.xml:61]: Thank you for the advice, could you please help me with how can I obtain/use the user id in this code? I'm not fluent with poms, always fighting with it. Thank you

File: tests/e2e/publish.sh (hunk 4 above)
Comment (Katka92) [reply tests/e2e/publish.sh]: Yeah, it's prepared for the https://github.com/eclipse/che/issues/18772 but I can comment it out.

File: tests/e2e/pom.xml [hunk 6]
```diff
@@ -0,0 +1,134 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2012-2018 Red Hat, Inc.
```
Comment (Katka92) [inline tests/e2e/pom.xml:4]: Can you please help me how to update this copyright to 2021? If I replace 2018 with 2021, I get error:
`[WARNING] Missing header in: /home/.../che/tests/e2e/pom.xml`

File: tests/e2e/pom.xml (hunk 2 above)
Comment (Katka92) [reply tests/e2e/pom.xml]: docker command replaced by bash, nodejs:10 replaced by nodejs:15

File: pom.xml [hunk 7]
```diff
@@ -31,6 +31,7 @@
         <module>infrastructures</module>
         <module>wsmaster/integration-tests</module>
         <module>assembly</module>
+        <module>tests/e2e</module>
```
Comment (sparkoo) [inline pom.xml:34]: does it mean it will run everytime I build che-server?

File: tests/e2e/pom.xml (hunk 6 above)
Comment (sparkoo) [reply tests/e2e/pom.xml:4]: we still live in 2018 :)

Comment (sparkoo) [reply tests/e2e/pom.xml:4]: https://github.com/eclipse/che/issues/18821

File: pom.xml (hunk 7 above)
Comment (Katka92) [reply pom.xml:34]: Every time you compile che-server, it should prepare tests. https://github.com/eclipse/che/pull/19524/files#diff-f8ed9675994a82a2c0a0d59b41c8395f50c117ccb62368aa8eae98b79ddedd66R53
Every time that `make-release.sh` is done (and so `maven deploy` is called) the test image should be pushed to npm registry. https://github.com/eclipse/che/pull/19524/files#diff-f8ed9675994a82a2c0a0d59b41c8395f50c117ccb62368aa8eae98b79ddedd66R92

I hope it make sense :) I've updated scripts today.

File: tests/e2e/pom.xml (hunk 6 above)
Comment (Katka92) [reply tests/e2e/pom.xml:4]: Ah, thanks for info :D

File: pom.xml (hunk 7 above)
Comment (sparkoo) [reply pom.xml:34]: thanks. how long does it takes to run? Can we exclude it in `fast` profile https://github.com/eclipse/che/blob/master/pom.xml#L1445 ?

Comment (sparkoo) [reply pom.xml:34]: or better don't run it by default and include it in some dedicated maven profile? To me it feels like it should not be run on every build.

**Comment by sparkoo** [comment]
@rhopp @Katka92 did you think about moving e2e testsuite to dedicated repository? Do the tests somehow depends on che-server codebase or does it have any other benefits to have it together?

AFAIK up to this point, the codebases were de-facto separated, only lived in the same repo. However this PR is merging the build systems and maven is not the right tool for nodejs project. Approach to just run docker from the maven is imho broken.

**Comment by Katka92** [comment]
We decided to not merge this PR. We agreed to move our test to our own repo and then implement this publishing mechanism. Thanks for your ideas @sparkoo 

This PR was also fixing the vulnerability dependencies, so I've moved that part to the new [PR](https://github.com/eclipse/che/pull/19615).

**Pull request closed without merging**